developer_guides/virtualisation/virtualisation.rst

```diff
@@ -0,0 +1,360 @@
+device ID from a range, specially donated by RedHat for this purpose.
+
+The SOF VirtIO driver on the guest registers support for that PCI device and
+its .probe() method is called.
+
+During the SOF VirtIO driver probing, the QEMU SOF code reports to the vhost
+driver, which VM image is being instantiated. This is used as a reliable guest
+identification and is needed for SOF guest audio topology selection.
+
+The SOF driver on the guest performs partial initialisation, it skips any
+steps, involving communicating with the actual DSP, e.g. sending the firmware,
+booting it, initialisation the tracing interface.
+
+The VirtIO driver on the guest uses a new SOF_IPC_TPLG_VFE_GET VirtIO-specific
+IPC message to request the topology from the host. The host reads the file,
+corresponding to the guest ID, obtained from the QEMU communication described
+above, from its filesystem, and sends it in multiple chunks to the guest in
+reply to that request.
+
+Guest topology files contain virtual DAI components of type "dai_out" and
+"dai_in." Their data tuples contain a new token of type SOF_TKN_COMP_REF, which
+contains an ID of the connection point, to which this pipeline should be
+attached.
+
+After obtaining the topology, the guest SOF instance uses a second new
+SOF_IPC_TPLG_VFE_COMP_ID VirtIO-specific IPC message to request its allocated
+component ID base. It then uses this base to assign IDs to all of its topology
+components. Those components are then sent to the DSP firmware. This way we
+make sure, that no two components sent to the DSP have the same ID. Note, that
+gaps in component IDs are allowed and don't have any side effects.
+
+The host maintains a list of guest component ID ranges to guarantee, that
+component IDs stay unique regardless of the order in which guests are brought up
+and down.
+
+Next the guest proceeds by sending the parsed topology over IPC to the host,
+which is then forwarded to the DSP firmware. The guest uses the .send_msg()
+callback to forward any IPC, including topology related, with no processing at
+all, to the host. The only exception is blocking IPCs when the guest is
+resuming, while the host doesn't need to be resumed (see section
+`3.3. Deactivation`_ on `reset count`_ for details).
+
+The host receives the guest topology over the IPC virtual queue. In most cases
+the host IPC handler just forwards IPCs to the DSP, but in some cases it has to
+handle or modify IPC contents locally. One such example is linking the guest and
+host topology fragments together. As mentioned above, guest topology files
+contain a new token, referencing connection point components in the host
+topology. This value is then assigned to the .ref_comp_id field of struct
+sof_ipc_comp_config during topology file parsing and sent by the guest to the
+host. The host then uses that value to modify SOF_IPC_TPLG_COMP_CONNECT
+connection IPC messages, involving connection points, from the guest before
+sending them to the DSP.
```

Comment:
maybe a stupid question, but couldn't we make sure that the topology sent by the host to the guest contains a set of unique component IDs, so that the guest never ever conflicts with other topologies, and that way we would remove the need for this translation layer?

We do want to uniquify components but do we really need the component base and translation back when the topology could implement unique IDs from day 1?

If you enable a range-based check in the host driver to make sure the IPCs sent by the host only target the components it instantiated and controls that would be a different story, but you didn't mention it here.

Reply:
@plbossart Component IDs are assigned during topology parsing, right? Guest topology fragments are parsed on guests, the host doesn't look into them at all (prior to sending). The host just reads in files from the filesystem and sends them out to guests. Besides component IDs aren't even present in topology files, there are probably even no fields for them there. So, we would have to change a lot for that - add processing on the host before sending, and change processing on the guest, because the data, it receives now isn't the standard topology data any more. So, no, don't think it's a good idea.

Topology IDs are unique, that's why we use their ranges. Translation is a different thing. It isn't used for all IDs. It's only used for connection points. Guests send their complete pipelines to the host. (See image). So, e.g. such a guest would send
```
[PCM]
[buffer]
[backend]
```
but the backend is "fake" / "virtual." What actually should happen is that the above PCM and buffer are connected to, say, a mixer on the host pipeline. So, when the guest sends the "backend" widget to the host, it will be dropped, instead of being sent to the DSP. And when the guest sends "connect buffer ID x to backend ID y" - *this it* where the translation happens. The host then replaces ID y with ID z of the mixer.

I don't have such a check yet, but that would also be a 1-2 line addition. All the data is already there. But again, that has nothing to do with connection point ID translation.

Reply:
Ah, I assumed the component IDs were part of the topology, but indeed if they are assigned dynamically while parsing then that's a wrong assumption. I may have learned something today :-)

@ranj063 can you confirm @lyakh 's point?